src/compiler/utilities.ts

```diff
@@ -4688,8 +4700,22 @@ namespace ts {
             getLineOfLocalPosition(sourceFile, pos1) === getLineOfLocalPosition(sourceFile, pos2);
     }
 
-    export function getStartPositionOfRange(range: TextRange, sourceFile: SourceFile) {
-        return positionIsSynthesized(range.pos) ? -1 : skipTrivia(sourceFile.text, range.pos);
+    export function getStartPositionOfRange(range: TextRange, sourceFile: SourceFile, includeComments: boolean) {
+        return positionIsSynthesized(range.pos) ? -1 : skipTrivia(sourceFile.text, range.pos, /*stopAfterLineBreak*/ false, includeComments);
+    }
+
+    export function getLinesBetweenPositionAndPrecedingNonWhitespaceCharacter(pos: number, sourceFile: SourceFile, includeComments?: boolean) {
+        const startPos = skipTrivia(sourceFile.text, pos, /*stopAfterLineBreak*/ false, includeComments);
+        const prevPos = getPreviousNonWhitespacePosition(startPos, sourceFile);
+        return getLineOfLocalPosition(sourceFile, startPos) - getLineOfLocalPosition(sourceFile, prevPos || 0);

```

Comment:
just curious -- how much work does `getLineOfLocalPosition` do?

Reply:
On the first call for a source file, it has to create a line map, which I suspect is pretty expensive. On subsequent calls, it does a binary search on the existing line map. When I was profiling to figure out where the last 1.5–2% of slowdown was coming from, I figured out that it was being called about 20% more than in master because I had forgotten to short circuit when positions are equal. (Now that you mention it, I forgot to do that in this function too, although this one only runs when emitting from TextChanges.) So, on most calls, it’s _fairly_ cheap, but expensive enough to be noticeable in a hot path like some of these other functions. (It’s worth noting that in a `tsc` compile, this is now called exactly as many times as before, as I simply replaced calls of “are these two positions on the same line” with calls of “what is the difference between the lines of these two positions” one-for-one.)

I’ll come back and add the optimization to this function too.

Reply:
It seems like we could possibly make `getLineOfLocalPosition` cheaper too, since it calls `computeLineAndCharacterOfPosition` which allocates an object, but we only use one property of the result.

There are also some possible minor optimizations we could consider, depending on whether they are valuable. For example, if we get the line for `startPos`, and we know that `prevPos` is actually less than `startPos`, then it stands to reason it must be on either the same or an earlier line, therefore the binary search of the line map does not necessarily need to search between `0` and `lineMap.length - 1` (inclusive), but instead can search between `0` and `startLine` (inclusive).

If you do such an optimization, I would recommend moving this logic to another function (i.e. `getLinesBetweenPositions`), since you use this logic in multiple places:

- https://github.com/microsoft/TypeScript/pull/36688/files#diff-942ae3cd2a8bbd85ed86a60cd7c43307R4687
- https://github.com/microsoft/TypeScript/pull/36688/files#diff-942ae3cd2a8bbd85ed86a60cd7c43307R4691

This would also help with keeping the shortcuts you have in the above two functions in a single place.